site/content/docs/how_do_compiler_functions_work.md

```diff
@@ -0,0 +1,170 @@
+     /1:N                   |                                                  |
+    /                       |                                                  v
+   |     +----------+       v        +----------------+                  +------------+
+   |     |  Engine  |--------------->|  ModuleEngine  |----------------->| CallEngine |
+   |     +----------+                +----------------+                  +------------+
+   |          |                               |                            |      |
+   .          |                               |                            |      |
+ foo.wasm --->|        .--------------------->|          '-----------------+      |
+              |       /                       |          |                        |
+              v      .                        v          v                        v
+      +--------------+      +-----------------------------------+            +----------+
+      | Machine Code |      |[(func_instance, machine_code),...]|            |Call Stack|
+      +--------------+      +-----------------------------------+            +----------+
+                                               ^                                  ^
+                                               |                                  |
+                                               |                                  |
+                                               +----------------------------------+
+                                                               |
+                                                               |
+                                                               |
+                                                        Function.Call()
+```
+
+## The behavior of machine code
+
+Go source can be compiled to invoke native library functions using CGO.
+However, [CGO is not GO][cgo-not-go]. To call native functions in pure Go, we
+need a different approach with unique constraints.
+
+For example, the generated machine code must not manipulate Goroutine
+(or system) stack. Otherwise, the Go runtime gets corrupted, which results in
+fatal execution errors. This means we cannot[^1] call Go functions (host
+functions) directly from machine code (compiled from wasm). This is routinely
+needed in WebAssembly, as system calls such as WASI are defined in Go, but
+invoked from Wasm.
+
+To handle this, we employ a "trampoline" strategy. Let's explain this with an
+example. `clock_time_get`, is a host function defined in Go, called from
+machine code compiled from guest wasm. Let's say the wasm function is named
+`getTime`.
+
+When `getTime` calls  `clock_time_get`, it actually exits execution first.
+wazero then calls the Go function mapped to `clock_time_get` like a usual Go
+program. Finally, wazero transfers control back to machine code again, resuming
+execution after the `clock_time_get` call instruction.
+
+TODO: add a diagram
```

Comment:
@mathetake can I trouble you with a diagram, using the function names above, not foo bar?

Reply:
the diagram in the next section is exactly for this.

Reply:
top-to-bottom editing problem :D